sentry-types/src/protocol/envelope.rs

```diff
@@ -236,6 +302,125 @@ impl Envelope {
+        let bytes = std::fs::read(path).map_err(|_| EnvelopeError::UnexpectedEof)?;
+        Envelope::from_slice(&bytes)
+    }
+
+    fn parse_header(slice: &[u8]) -> Result<(EnvelopeHeader, usize), EnvelopeError> {
+        let mut stream = serde_json::Deserializer::from_slice(slice).into_iter();
+
+        let header: EnvelopeHeader = match stream.next() {
+            None => return Err(EnvelopeError::MissingHeader),
+            Some(Err(error)) => return Err(EnvelopeError::InvalidHeader(error)),
+            Some(Ok(header)) => header,
+        };
+
+        // Each header is terminated by a UNIX newline.
+        Self::require_termination(slice, stream.byte_offset())?;
+
+        Ok((header, stream.byte_offset() + 1))
+    }
+
+    fn parse_items(slice: &[u8], mut offset: usize) -> Result<Vec<EnvelopeItem>, EnvelopeError> {
+        let mut items = Vec::new();
+
+        while offset < slice.len() {
+            let bytes = slice
+                .get(offset..)
+                .ok_or(EnvelopeError::MissingItemHeader)?;
+            let (item, item_size) = Self::parse_item(bytes)?;
+            offset += item_size;
+            items.push(item);
+        }
+
+        Ok(items)
+    }
+
+    fn parse_item(slice: &[u8]) -> Result<(EnvelopeItem, usize), EnvelopeError> {
+        let mut stream = serde_json::Deserializer::from_slice(slice).into_iter();
+
+        let header: EnvelopeItemHeader = match stream.next() {
+            None => return Err(EnvelopeError::UnexpectedEof),
+            Some(Err(error)) => return Err(EnvelopeError::InvalidItemHeader(error)),
+            Some(Ok(header)) => header,
+        };
+
+        // Each header is terminated by a UNIX newline.
+        let header_end = stream.byte_offset();
+        Self::require_termination(slice, header_end)?;
+
+        // The last header does not require a trailing newline, so `payload_start` may point
+        // past the end of the buffer.
+        let payload_start = std::cmp::min(header_end + 1, slice.len());
+        let payload_end = match header.length {
+            Some(len) => {
+                let payload_end = payload_start + len as usize;
+                if slice.len() < payload_end {
+                    return Err(EnvelopeError::UnexpectedEof);
+                }
+
+                // Each payload is terminated by a UNIX newline.
+                Self::require_termination(slice, payload_end)?;
+                payload_end
+            }
+            None => match slice[payload_start..].iter().position(|b| *b == b'\n') {
+                Some(relative_end) => payload_start + relative_end,
+                None => slice.len(),
+            },
+        };
+
+        let payload = slice.get(payload_start..payload_end).unwrap();
+
+        let item = match header.r#type {
+            EnvelopeItemType::Event => EnvelopeItem::Event(
+                serde_json::from_slice(payload).map_err(EnvelopeError::InvalidItemPayload)?,
```

Comment:
This duplicated `map_err` is a bit annoying.

We could potentially make this `serde_json::from_slice(payload).map(EnvelopeItem::Event)`, and an explicit `Ok(EnvelopeItem::Attachment(…))` below.
That way we can move the `.map_err(…)?` outside the match. That can potentially improve readability.

Reply:
So much better now :)